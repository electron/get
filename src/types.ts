import { Downloader } from './Downloader';

export { Downloader };

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type DownloadOptions = any;

export interface MirrorOptions {
  /**
   * DEPRECATED - see nightlyMirror.
   */
  nightly_mirror?: string;
  /**
   * The Electron nightly-specific mirror URL.
   */
  nightlyMirror?: string;
  /**
   * The base URL of the mirror to download from,
   * e.g https://github.com/electron/electron/releases/download
   */
  mirror?: string;
  /**
   * The name of the directory to download from,
   * often scoped by version number e.g 'v4.0.4'
   */
  customDir?: string;
  /**
   * The name of the asset to download,
   * e.g 'electron-v4.0.4-linux-x64.zip'
   */
  customFilename?: string;
  /**
   * The version of the asset to download,
   * e.g '4.0.4'
   */
  customVersion?: string;
  /**
   * A function allowing customization of the url returned
   * from getArtifactRemoteURL().
   */
  resolveAssetURL?: (opts: DownloadOptions) => Promise<string>;
}

export interface ElectronDownloadRequest {
  /**
   * The version of Electron associated with the artifact.
   */
  version: string;
  /**
   * The type of artifact. For example:
   * * `electron`
   * * `ffmpeg`
   */
  artifactName: string;
}

export interface ElectronDownloadRequestOptions {
  /**
   * Whether to download an artifact regardless of whether it's in the cache directory.
   *
   * Defaults to `false`.
   */
  force?: boolean;
  /**
   * When set to `true`, disables checking that the artifact download completed successfully
   * with the correct payload.
   *
   * Defaults to `false`.
   */
  unsafelyDisableChecksums?: boolean;
  /**
   * Provides checksums for the artifact as strings.
   * Can be used if you already know the checksums of the Electron artifact
   * you are downloading and want to skip the checksum file download
   * without skipping the checksum validation.
   *
   * This should be an object whose keys are the file names of the artifacts and
   * the values are their respective SHA256 checksums.
   */
  checksums?: Record<string, string>;
  /**
   * The directory that caches Electron artifact downloads.
   *
   * The default value is dependent upon the host platform:
   *
   * * Linux: `$XDG_CACHE_HOME` or `~/.cache/electron/`
   * * MacOS: `~/Library/Caches/electron/`
   * * Windows: `%LOCALAPPDATA%/electron/Cache` or `~/AppData/Local/electron/Cache/`
   */
  cacheRoot?: string;
  /**
   * Options passed to the downloader module.
   */
  downloadOptions?: DownloadOptions;
  /**
   * Options related to specifying an artifact mirror.
   */
  mirrorOptions?: MirrorOptions;
  /**
   * The custom {@link Downloader} class used to download artifacts. Defaults to the
   * built-in {@link GotDownloader}.
   */
  downloader?: Downloader<DownloadOptions>;
  /**
   * A temporary directory for downloads.
   * It is used before artifacts are put into cache.
   */
  tempDirectory?: string;
}

export type ElectronPlatformArtifactDetails = {
  /**
   * The target artifact platform. These are Node-style platform names, for example:
   * * `win32`
   * * `darwin`
   * * `linux`
   */
  platform: string;
  /**
   * The target artifact architecture. These are Node-style architecture names, for example:
   * * `ia32`
   * * `x64`
   * * `armv7l`
   */
  arch: string;
  artifactSuffix?: string;
  isGeneric?: false;
} & ElectronDownloadRequest &
  ElectronDownloadRequestOptions;

export type ElectronGenericArtifactDetails = {
  isGeneric: true;
} & ElectronDownloadRequest &
  ElectronDownloadRequestOptions;

export type ElectronArtifactDetails =
  | ElectronPlatformArtifactDetails
  | ElectronGenericArtifactDetails;

export type Omit<T, K> = Pick<T, Exclude<keyof T, K>>;

export type ElectronPlatformArtifactDetailsWithDefaults =
  | (Omit<ElectronPlatformArtifactDetails, 'platform' | 'arch'> & {
      platform?: string;
      arch?: string;
    })
  | ElectronGenericArtifactDetails;
