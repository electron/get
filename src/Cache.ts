import debug from 'debug';
import envPaths from 'env-paths';
import * as fs from 'fs-extra';
import * as path from 'path';
import * as url from 'url';
import * as crypto from 'crypto';

const d = debug('@electron/get:cache');

const defaultCacheRoot = envPaths('electron', {
  suffix: '',
}).cache;

export class Cache {
  constructor(private cacheRoot = defaultCacheRoot) {}

  public static getCacheDirectory(downloadUrl: string): string {
    const parsedDownloadUrl = url.parse(downloadUrl);
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const { search, hash, pathname, ...rest } = parsedDownloadUrl;
    const strippedUrl = url.format({ ...rest, pathname: path.dirname(pathname || 'electron') });

    return crypto.createHash('sha256').update(strippedUrl).digest('hex');
  }

  public getCachePath(downloadUrl: string, fileName: string): string {
    return path.resolve(this.cacheRoot, Cache.getCacheDirectory(downloadUrl), fileName);
  }

  public async getPathForFileInCache(url: string, fileName: string): Promise<string | null> {
    const cachePath = this.getCachePath(url, fileName);
    if (await fs.pathExists(cachePath)) {
      return cachePath;
    }

    return null;
  }

  public async putFileInCache(url: string, currentPath: string, fileName: string): Promise<string> {
    const cachePath = this.getCachePath(url, fileName);
    d(`Moving ${currentPath} to ${cachePath}`);
    if (await fs.pathExists(cachePath)) {
      d('* Replacing existing file');
      await fs.remove(cachePath);
    }

    await fs.move(currentPath, cachePath);

    return cachePath;
  }
}
