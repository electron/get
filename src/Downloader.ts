export interface Downloader<T> {
  download(url: string, targetFilePath: string, options: T): Promise<void>;
}
