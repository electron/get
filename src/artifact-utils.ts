import { ElectronArtifactDetails, MirrorOptions } from './types.js';
import { ensureIsTruthyString, normalizeVersion } from './utils.js';

const BASE_URL = 'https://github.com/electron/electron/releases/download/';
const NIGHTLY_BASE_URL = 'https://github.com/electron/nightlies/releases/download/';

export function getArtifactFileName(details: ElectronArtifactDetails): string {
  ensureIsTruthyString(details, 'artifactName');

  if (details.isGeneric) {
    return details.artifactName;
  }

  ensureIsTruthyString(details, 'arch');
  ensureIsTruthyString(details, 'platform');
  ensureIsTruthyString(details, 'version');

  return `${[
    details.artifactName,
    details.version,
    details.platform,
    details.arch,
    ...(details.artifactSuffix ? [details.artifactSuffix] : []),
  ].join('-')}.zip`;
}

function mirrorVar(
  name: keyof Omit<MirrorOptions, 'resolveAssetURL'>,
  options: MirrorOptions,
  defaultValue: string,
): string {
  // Convert camelCase to camel_case for env var reading
  const snakeName = name.replace(/([a-z])([A-Z])/g, (_, a, b) => `${a}_${b}`).toLowerCase();

  return (
    // .npmrc
    process.env[`npm_config_electron_${name.toLowerCase()}`] ||
    process.env[`NPM_CONFIG_ELECTRON_${snakeName.toUpperCase()}`] ||
    process.env[`npm_config_electron_${snakeName}`] ||
    // package.json
    process.env[`npm_package_config_electron_${name}`] ||
    process.env[`npm_package_config_electron_${snakeName.toLowerCase()}`] ||
    // env
    process.env[`ELECTRON_${snakeName.toUpperCase()}`] ||
    options[name] ||
    defaultValue
  );
}

export async function getArtifactRemoteURL(details: ElectronArtifactDetails): Promise<string> {
  const opts: MirrorOptions = details.mirrorOptions || {};
  let base = mirrorVar('mirror', opts, BASE_URL);
  if (details.version.includes('nightly')) {
    base = mirrorVar('nightlyMirror', opts, NIGHTLY_BASE_URL);
  }
  const path = mirrorVar('customDir', opts, details.version).replace(
    '{{ version }}',
    details.version.replace(/^v/, ''),
  );
  const file = mirrorVar('customFilename', opts, getArtifactFileName(details));

  // Allow customized download URL resolution.
  if (opts.resolveAssetURL) {
    const url = await opts.resolveAssetURL(details);
    return url;
  }

  return `${base}${path}/${file}`;
}

export function getArtifactVersion(details: ElectronArtifactDetails): string {
  return normalizeVersion(mirrorVar('customVersion', details.mirrorOptions || {}, details.version));
}
