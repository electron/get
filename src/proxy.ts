import * as debug from 'debug';
import { getEnv, setEnv } from './utils';

const d = debug('@electron/get:proxy');

/**
 * Initializes a third-party proxy module for HTTP(S) requests.
 */
export function initializeProxy(): void {
  try {
    // Code originally from https://github.com/yeoman/yo/blob/b2eea87e/lib/cli.js#L19-L28
    const MAJOR_NODEJS_VERSION = parseInt(process.version.slice(1).split('.')[0], 10);

    if (MAJOR_NODEJS_VERSION >= 10) {
      // See: https://github.com/electron/get/pull/214#discussion_r798845713
      const env = getEnv('GLOBAL_AGENT_');

      setEnv('GLOBAL_AGENT_HTTP_PROXY', env('HTTP_PROXY'));
      setEnv('GLOBAL_AGENT_HTTPS_PROXY', env('HTTPS_PROXY'));
      setEnv('GLOBAL_AGENT_NO_PROXY', env('NO_PROXY'));

      // `global-agent` works with Node.js v10 and above.
      require('global-agent').bootstrap();
    } else {
      // `global-tunnel-ng` works with Node.js v10 and below.
      require('global-tunnel-ng').initialize();
    }
  } catch (e) {
    d('Could not load either proxy modules, built-in proxy support not available:', e);
  }
}
