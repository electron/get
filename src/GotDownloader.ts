/* eslint-disable */
// @ts-ignore - require(esm) supported in Node 22.12
import got, { HTTPError, Progress as GotProgress, Options as GotOptions } from 'got';
/* eslint-enable */

import fs from 'node:fs';
import path from 'node:path';
import ProgressBar from 'progress';

import { Downloader } from './Downloader';

const PROGRESS_BAR_DELAY_IN_SECONDS = 30;

/**
 * Options for the default [`got`](https://github.com/sindresorhus/got) Downloader implementation.
 *
 * @category Downloader
 * @see {@link https://github.com/sindresorhus/got/tree/v11.8.5?tab=readme-ov-file#options | `got#options`} for possible keys/values.
 */
export type GotDownloaderOptions = GotOptions & { isStream?: true } & {
  /**
   * if defined, triggers every time `got`'s
   * {@link https://github.com/sindresorhus/got/tree/v11.8.5?tab=readme-ov-file#downloadprogress | `downloadProgress``} event callback is triggered.
   */
  getProgressCallback?: (progress: GotProgress) => Promise<void>;
  /**
   * if `true`, disables the console progress bar (setting the `ELECTRON_GET_NO_PROGRESS`
   * environment variable to a non-empty value also does this).
   */
  quiet?: boolean;
};

/**
 * Default {@link Downloader} implemented with {@link https://npmjs.com/package/got | `got`}.
 * @category Downloader
 */
export class GotDownloader implements Downloader<GotDownloaderOptions> {
  async download(
    url: string,
    targetFilePath: string,
    options?: Partial<GotDownloaderOptions>,
  ): Promise<void> {
    if (!options) {
      options = {};
    }
    const { quiet, getProgressCallback, ...gotOptions } = options;
    let downloadCompleted = false;
    let bar: ProgressBar | undefined;
    let progressPercent: number;
    let timeout: NodeJS.Timeout | undefined = undefined;
    await fs.promises.mkdir(path.dirname(targetFilePath), { recursive: true });
    const writeStream = fs.createWriteStream(targetFilePath);

    if (!quiet || !process.env.ELECTRON_GET_NO_PROGRESS) {
      const start = new Date();
      timeout = setTimeout(() => {
        if (!downloadCompleted) {
          bar = new ProgressBar(
            `Downloading ${path.basename(url)}: [:bar] :percent ETA: :eta seconds `,
            {
              curr: progressPercent,
              total: 100,
            },
          );
          // https://github.com/visionmedia/node-progress/issues/159
          // eslint-disable-next-line @typescript-eslint/no-explicit-any
          (bar as any).start = start;
        }
      }, PROGRESS_BAR_DELAY_IN_SECONDS * 1000);
    }
    await new Promise<void>((resolve, reject) => {
      const downloadStream = got.stream(url, gotOptions);
      downloadStream.on('downloadProgress', async (progress) => {
        progressPercent = progress.percent;
        if (bar) {
          bar.update(progress.percent);
        }
        if (getProgressCallback) {
          await getProgressCallback(progress);
        }
      });
      downloadStream.on('error', (error) => {
        if (error instanceof HTTPError && error.response.statusCode === 404) {
          error.message += ` for ${error.response.url}`;
        }
        if (writeStream.destroy) {
          writeStream.destroy(error);
        }

        reject(error);
      });
      writeStream.on('error', (error) => reject(error));
      writeStream.on('close', () => resolve());

      downloadStream.pipe(writeStream);
    });

    downloadCompleted = true;
    if (timeout) {
      clearTimeout(timeout);
    }
  }
}
