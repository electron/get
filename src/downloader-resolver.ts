import { DownloadOptions } from './types.js';
import { Downloader } from './Downloader.js';

// TODO: Resolve the downloader or default to GotDownloader
// Current thoughts are a dot-file traversal for something like
// ".electron.downloader" which would be a text file with the name of the
// npm module to import() and use as the downloader
import { GotDownloader } from './GotDownloader.js';

export async function getDownloaderForSystem(): Promise<Downloader<DownloadOptions>> {
  return new GotDownloader();
}
