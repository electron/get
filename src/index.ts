import debug from 'debug';
import * as path from 'path';
import * as sumchecker from 'sumchecker';

import { getArtifactFileName, getArtifactRemoteURL } from './artifact-utils';
import {
  ElectronArtifactDetails,
  ElectronDownloadRequestOptions,
  ElectronPlatformArtifactDetailsWithDefaults,
} from './types';
import { Cache } from './Cache';
import { getDownloaderForSystem } from './downloader-resolver';
import { initializeProxy } from './proxy';
import {
  withTempDirectoryIn,
  normalizeVersion,
  getHostArch,
  getNodeArch,
  ensureIsTruthyString,
  isOfficialLinuxIA32Download,
} from './utils';

export { getHostArch } from './utils';
export { initializeProxy } from './proxy';
export * from './types';

const d = debug('@electron/get:index');

if (process.env.ELECTRON_GET_USE_PROXY) {
  initializeProxy();
}

/**
 * Downloads a specific version of Electron and returns an absolute path to a
 * ZIP file.
 *
 * @param version - The version of Electron you want to download
 */
export function download(
  version: string,
  options?: ElectronDownloadRequestOptions,
): Promise<string> {
  return downloadArtifact({
    ...options,
    version,
    platform: process.platform,
    arch: process.arch,
    artifactName: 'electron',
  });
}

/**
 * Downloads an artifact from an Electron release and returns an absolute path
 * to the downloaded file.
 *
 * @param artifactDetails - The information required to download the artifact
 */
export async function downloadArtifact(
  _artifactDetails: ElectronPlatformArtifactDetailsWithDefaults,
): Promise<string> {
  const artifactDetails: ElectronArtifactDetails = _artifactDetails.isGeneric
    ? {
        ..._artifactDetails,
      }
    : {
        platform: process.platform,
        arch: process.arch,
        ..._artifactDetails,
      };
  ensureIsTruthyString(artifactDetails, 'version');
  artifactDetails.version = normalizeVersion(artifactDetails.version);
  if (artifactDetails.hasOwnProperty('arch')) {
    (artifactDetails as any).arch = getNodeArch((artifactDetails as any).arch);
  }

  const fileName = getArtifactFileName(artifactDetails);
  const url = getArtifactRemoteURL(artifactDetails);
  const cache = new Cache(artifactDetails.cacheRoot);

  // Do not check if the file exists in the cache when force === true
  if (!artifactDetails.force) {
    d(`Checking the cache (${artifactDetails.cacheRoot}) for ${fileName} (${url})`);
    const cachedPath = await cache.getPathForFileInCache(url, fileName);

    if (cachedPath === null) {
      d('Cache miss');
    } else {
      d('Cache hit');
      return cachedPath;
    }
  }

  if (
    !artifactDetails.isGeneric &&
    isOfficialLinuxIA32Download(
      artifactDetails.platform,
      artifactDetails.arch,
      artifactDetails.version,
      artifactDetails.mirrorOptions,
    )
  ) {
    console.warn('Official Linux/ia32 support is deprecated.');
    console.warn('For more info: https://electronjs.org/blog/linux-32bit-support');
  }

  return await withTempDirectoryIn(artifactDetails.tempDirectory, async tempFolder => {
    const tempDownloadPath = path.resolve(tempFolder, getArtifactFileName(artifactDetails));

    const downloader = artifactDetails.downloader || (await getDownloaderForSystem());
    d(
      `Downloading ${url} to ${tempDownloadPath} with options: ${JSON.stringify(
        artifactDetails.downloadOptions,
      )}`,
    );
    await downloader.download(url, tempDownloadPath, artifactDetails.downloadOptions);

    // Don't try to verify the hash of the hash file itself
    if (
      !artifactDetails.artifactName.startsWith('SHASUMS256') &&
      !artifactDetails.unsafelyDisableChecksums
    ) {
      const shasumPath = await downloadArtifact({
        isGeneric: true,
        version: artifactDetails.version,
        artifactName: 'SHASUMS256.txt',
        force: artifactDetails.force,
        downloadOptions: artifactDetails.downloadOptions,
        cacheRoot: artifactDetails.cacheRoot,
        downloader: artifactDetails.downloader,
        mirrorOptions: artifactDetails.mirrorOptions,
      });

      await sumchecker('sha256', shasumPath, path.dirname(tempDownloadPath), [
        path.basename(tempDownloadPath),
      ]);
    }

    return await cache.putFileInCache(url, tempDownloadPath, fileName);
  });
}
