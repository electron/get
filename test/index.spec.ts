import * as crypto from 'crypto';
import * as fs from 'fs-extra';
import * as os from 'os';
import * as path from 'path';

import { FixtureDownloader } from './FixtureDownloader';
import { download, downloadArtifact } from '../src';

describe('Public API', () => {
  const downloader = new FixtureDownloader();

  let cacheRoot: string;
  beforeEach(async () => {
    cacheRoot = await fs.mkdtemp(path.resolve(os.tmpdir(), 'electron-download-spec-'));
  });

  afterEach(async () => {
    await fs.remove(cacheRoot);
  });

  describe('download()', () => {
    it('should return a valid path to a downloaded zip file', async () => {
      const zipPath = await download('2.0.10', {
        cacheRoot,
        downloader,
      });
      expect(typeof zipPath).toEqual('string');
      expect(await fs.pathExists(zipPath)).toEqual(true);
      expect(path.extname(zipPath)).toEqual('.zip');
    });

    it('should return a valid path to a downloaded zip file for nightly releases', async () => {
      const zipPath = await download('6.0.0-nightly.20190213', {
        cacheRoot,
        downloader,
      });
      expect(typeof zipPath).toEqual('string');
      expect(await fs.pathExists(zipPath)).toEqual(true);
      expect(path.extname(zipPath)).toEqual('.zip');
    });

    it('should not redownload when force=false', async () => {
      const zipPath = await download('2.0.9', {
        cacheRoot,
        downloader,
        force: false,
      });
      await fs.writeFile(zipPath, 'bad content');
      const zipPath2 = await download('2.0.9', {
        cacheRoot,
        downloader,
        force: false,
      });
      expect(zipPath).toEqual(zipPath2);
      expect(await fs.readFile(zipPath, 'utf8')).toEqual('bad content');
    });

    it('should redownload when force=true', async () => {
      const zipPath = await download('2.0.9', {
        cacheRoot,
        downloader,
        force: true,
      });
      const hash = crypto
        .createHash('sha256')
        .update(await fs.readFile(zipPath))
        .digest('hex');
      await fs.writeFile(zipPath, 'bad content');
      const zipPath2 = await download('2.0.9', {
        cacheRoot,
        downloader,
        force: true,
      });
      expect(zipPath).toEqual(zipPath2);
      const hash2 = crypto
        .createHash('sha256')
        .update(await fs.readFile(zipPath2))
        .digest('hex');
      expect(hash).toEqual(hash2);
    });

    it('should accept a custom downloader', async () => {
      const zipPath = await download('2.0.3', {
        cacheRoot,
        unsafelyDisableChecksums: true,
        downloader: {
          async download(url: string, targetPath: string) {
            expect(
              url.replace(process.platform, 'platform').replace(process.arch, 'arch'),
            ).toMatchSnapshot();
            await fs.writeFile(targetPath, 'faked from downloader');
          },
        },
      });
      expect(await fs.readFile(zipPath, 'utf8')).toEqual('faked from downloader');
    });

    it('should pass download options to a custom downloader', async () => {
      const downloadOpts = {
        magic: 'option',
        trick: 'shot',
      };
      await download('2.0.3', {
        cacheRoot,
        unsafelyDisableChecksums: true,
        downloader: {
          async download(url: string, targetPath: string, opts?: any) {
            expect(opts).toStrictEqual(downloadOpts);
            await fs.writeFile(targetPath, 'file');
          },
        },
        downloadOptions: downloadOpts,
      });
    });
  });

  describe('downloadArtifact()', () => {
    it('should work for electron.d.ts', async () => {
      const dtsPath = await downloadArtifact({
        cacheRoot,
        downloader,
        isGeneric: true,
        version: '2.0.9',
        artifactName: 'electron.d.ts',
      });
      expect(await fs.pathExists(dtsPath)).toEqual(true);
      expect(path.basename(dtsPath)).toEqual('electron.d.ts');
      expect(await fs.readFile(dtsPath, 'utf8')).toContain('declare namespace Electron');
    });

    it('should work with the default platform/arch', async () => {
      await downloadArtifact({
        downloader,
        version: '2.0.3',
        artifactName: 'electron',
      });
    });

    it('should download linux/armv7l when linux/arm is passed as platform/arch', async () => {
      const zipPath = await downloadArtifact({
        cacheRoot,
        downloader,
        artifactName: 'electron',
        version: '2.0.9',
        platform: 'linux',
        arch: 'arm',
      });
      expect(path.basename(zipPath)).toMatchInlineSnapshot(`"electron-v2.0.9-linux-armv7l.zip"`);
    });

    it('should work for chromedriver', async () => {
      const driverPath = await downloadArtifact({
        cacheRoot,
        downloader,
        version: '2.0.9',
        artifactName: 'chromedriver',
        platform: 'darwin',
        arch: 'x64',
      });
      expect(await fs.pathExists(driverPath)).toEqual(true);
      expect(path.basename(driverPath)).toMatchInlineSnapshot(
        `"chromedriver-v2.0.9-darwin-x64.zip"`,
      );
      expect(path.extname(driverPath)).toEqual('.zip');
    });
  });
});
