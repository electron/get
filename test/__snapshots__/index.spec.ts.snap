// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Public API > download() > should accept a custom downloader 1`] = `"https://github.com/electron/electron/releases/download/v2.0.3/electron-v2.0.3-platform-arch.zip"`;
